File: app/src/main/java/net/juniet/ndx/MainActivity.java
package net.juniet.ndx;

import android.content.SharedPreferences;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.text.method.ScrollingMovementMethod;
import android.view.View;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {
    public static final String PREFS_NAME = "preferences";
    private final String VERSION = "0.3";
    private int rollsCount = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        SharedPreferences settings = getSharedPreferences(PREFS_NAME, 0);

        TextView textView = (TextView) findViewById(R.id.rollResults);
        textView.setMovementMethod(new ScrollingMovementMethod());
        textView.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                clearRollLogs((TextView) v);
                return true;
            }
        });

        TableLayout table = (TableLayout) findViewById(R.id.tableLayout);
        for (int i = 0; i < table.getChildCount(); ++i) {
            TableRow row = (TableRow) table.getChildAt(i);
            for (int j = 0; j < row.getChildCount(); ++j) {
                DiceButton bt = (DiceButton) row.getChildAt(j);
                String label = settings.getString("label_" + bt.getTag(), bt.getTag().toString());
                bt.setText(label);
                String formula = settings.getString("formula_" + bt.getTag(), bt.getRollModel().toString());
                bt.setRollModel(new DiceRollModel(formula));
            }
        }

        if (settings.getBoolean("showHelp", true)) {
            Toast.makeText(getApplicationContext(), getString(R.string.help), Toast.LENGTH_LONG).show();
        }
    }

    @Override
    protected void onStop(){
        super.onStop();

        SharedPreferences settings = getSharedPreferences(PREFS_NAME, 0);
        SharedPreferences.Editor editor = settings.edit();
        editor.putString("version", VERSION);
        editor.putBoolean("showHelp", false);

        TableLayout table = (TableLayout) findViewById(R.id.tableLayout);
        for (int i = 0; i < table.getChildCount(); ++i) {
            TableRow row = (TableRow) table.getChildAt(i);
            for (int j = 0; j < row.getChildCount(); ++j) {
                DiceButton bt = (DiceButton) row.getChildAt(j);
                editor.putString("label_" + bt.getTag(), bt.getText().toString());
                editor.putString("formula_" + bt.getTag(), bt.getRollModel().toString());
            }
        }

        editor.apply();
    }

    public void rollDice(View view) {
        DiceRollModel model = ((DiceButton)view).getRollModel();
        DiceRollResult result = model.getRollResult();
        addRollLog(model, result);
    }

    private void addRollLog(DiceRollModel model, DiceRollResult result) {
        StringBuilder sb = new StringBuilder();
        sb.append("\n\n#").append(++rollsCount);
        sb.append(": ").append(model.toString());
        sb.append(" = ").append(result.toString());

        final TextView text = (TextView) findViewById(R.id.rollResults);
        text.append(sb.toString());
    }

    private void clearRollLogs(TextView text) {
        text.setText(null);
        rollsCount = 0;
    }
}

File: app/src/main/java/net/juniet/ndx/MainActivity.kt
package net.juniet.ndx

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.text.method.ScrollingMovementMethod
import android.view.View
import android.widget.TableLayout
import android.widget.TableRow
import android.widget.TextView
import android.widget.Toast

class MainActivity : AppCompatActivity() {
    private val PREFS_NAME = "preferences"
    private val PREFS_VERSION = "0.3"
    private var rollsCount = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val settings = getSharedPreferences(PREFS_NAME, 0)

        val textView = findViewById(R.id.rollResults) as TextView
        textView.movementMethod = ScrollingMovementMethod()
        textView.setOnLongClickListener { v ->
            clearRollLogs(v as TextView)
            true
        }

        val table = findViewById(R.id.tableLayout) as TableLayout
        for (i in 0..table.childCount - 1) {
            val row = table.getChildAt(i) as TableRow
            for (j in 0..row.childCount - 1) {
                val bt = row.getChildAt(j) as DiceButton
                val label = settings.getString("label_" + bt.tag, bt.tag.toString())
                bt.text = label
                val formula = settings.getString("formula_" + bt.tag, bt.rollModel.toString())
                bt.rollModel = DiceRollModel(formula)
            }
        }

        if (settings.getBoolean("showHelp", true)) {
            Toast.makeText(applicationContext, getString(R.string.help), Toast.LENGTH_LONG).show()
        }
    }

    override fun onStop() {
        super.onStop()

        val settings = getSharedPreferences(PREFS_NAME, 0)
        val editor = settings.edit()
        editor.putString("version", PREFS_VERSION)
        editor.putBoolean("showHelp", false)

        val table = findViewById(R.id.tableLayout) as TableLayout
        for (i in 0..table.childCount - 1) {
            val row = table.getChildAt(i) as TableRow
            for (j in 0..row.childCount - 1) {
                val bt = row.getChildAt(j) as DiceButton
                editor.putString("label_" + bt.tag, bt.text.toString())
                editor.putString("formula_" + bt.tag, bt.rollModel.toString())
            }
        }

        editor.apply()
    }

    fun rollDice(view: View) {
        val model = (view as DiceButton).rollModel
        val result = model.rollResult
        addRollLog(model, result)
    }

    private fun addRollLog(model: DiceRollModel, result: DiceRollResult) {
        val sb = StringBuilder()
        sb.append("\n\n#").append(++rollsCount)
        sb.append(": ").append(model.toString())
        sb.append(" = ").append(result.toString())

        val text = findViewById(R.id.rollResults) as TextView
        text.append(sb.toString())
    }

    private fun clearRollLogs(text: TextView) {
        text.text = null
        rollsCount = 0
    }
}
